### PR TITLE
How do you solve a problem like ad-blocking?

### PR DESCRIPTION
**Pull request by seanmtracey**
Initial ad-block detection and alternate content insertion

Handle ad-blocking

Initial implementation of adding ad-block detection to our sites and replacing blocked ads with alternative content.

Removed ad-block detection code

Modularise anti-blocker approach code

add bart script

Beginnings of a controllable UI

Added UI controls to allow people to select the different approaches

Changed the structure for the different approaches.

add ad-revenue filter

Handle elements in paragraphs one level deep

Chalkboard ad-block-handling demo. Not active atm, as content is not put back in the page correctly

Included ad-revenue in approaches array. Slightly adjusted the module structure

Slight tweak of storage logic

Added let's talk demo.

Removed break; in switch statement. Pulled latest master and resolved Bower conflicts

add articles related to ad blocking to body

initial ad-block stuff

Initial ad-block detection and alternate content insertion

Handle ad-blocking

Initial implementation of adding ad-block detection to our sites and replacing blocked ads with alternative content.

Removed ad-block detection code

Modularise anti-blocker approach code

Beginnings of a controllable UI

Added UI controls to allow people to select the different approaches

Slight tweak of storage logic

Added let's talk demo.

Removed break; in switch statement. Pulled latest master and resolved Bower conflicts

add articles related to ad blocking to body

linting things

Rebased master onto this

Linting

Added link to form for feedback

Removing duplicate variable definition. Removing labs markup from ad spaces

### CODE REVIEW

File: bower.json [hunk 1]
```diff
@@ -25,6 +25,11 @@
     "o-ft-icons": "next",
     "o-fonts": "^2.1.2",
     "o-typography": "^3.2.0",
-		"o-hoverable": "^3.0.0"
+    "o-hoverable": "^3.0.0",
```
Comment (andygnewman) [inline bower.json]: We shouldn't need all these resolutions; master branch only has;

```
    "o-ft-icons": "next",
    "o-fonts": "^2.1.2",
    "o-typography": "^3.2.0",
    "o-hoverable": "^3.0.0"
```
and we're not requiring anything that should need resolutions, so suggest that we revert to what master has.

File: client/components/ftlabsadblockerhandling/ad-blocking-articles.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+var fetchres = require('fetchres');
+var oDate = require('o-date');
+
+module.exports = {
+
+	name: 'Ad Blocking Articles',
+	description: 'What the FT has written related to ad-blocking',
+	run: () => {
+		console.log('running ad-blocking articles');
```
Comment (andygnewman) [inline client/components/ftlabsadblockerhandling/ad-blocking-articles.js]: we can probably remove the console.logs

File: client/components/ftlabsadblockerhandling/chalkboard.js [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+		if(amountToType === 0){
+			h.show();
+		} else {
+			document.querySelector('.i-will-not .amt').textContent = amountToType;
+			document.querySelector('.i-will-not input').value = "";
+		}
+	}
+}
+
+module.exports = {
```
Comment (andygnewman) [inline client/components/ftlabsadblockerhandling/chalkboard.js]: Is this one actually being used - not referenced in the main.js?

File: client/components/ftlabsadblockerhandling/lib/content-hider.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
+let contentElement = undefined
```
Comment (andygnewman) [inline client/components/ftlabsadblockerhandling/lib/content-hider.js]: why not just use `.style.display = 'none' / 'block'`

File: client/components/ftlabsadblockerhandling/main.scss [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+.ftlabs-ad-block-handling-ui .controls {
+	text-align: right;
+	margin-bottom: 5px;
+}
+
+.ftlabs-ad-block-handling-ui button {
+	padding-left: 0;
+		background-position: 50%;
```
Comment (andygnewman) [inline client/components/ftlabsadblockerhandling/main.scss]: any reason for this indenting?

File: client/components/ftlabsadblockerhandling/chalkboard.js (hunk 3 above)
Comment (seanmtracey) [inline client/components/ftlabsadblockerhandling/chalkboard.js]: Yeah, this shouldn't be in here. It was an idea that never got finished. Will remove.

File: client/components/ftlabsadblockerhandling/no-vowels.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+	name : "no-vowels",
+	description : "No Ads. No Vowels",
+	run : function(){
+
+		const contentPTags = document.querySelectorAll('.article__body.n-content-body p');
```
Comment (andygnewman) [inline client/components/ftlabsadblockerhandling/no-vowels.js]: selector only needs `'.article__body p'`

File: client/components/ftlabsadblockerhandling/lib/content-hider.js (hunk 4 above)
Comment (seanmtracey) [inline client/components/ftlabsadblockerhandling/lib/content-hider.js]: I tried to be clever with the hiding of the content.

File: client/main.js [hunk 7]
```diff
@@ -35,7 +42,7 @@ require('next-js-setup').bootstrap(({flags}) => {
 	flags.get('saveForLater') && clientOpts.push({relationship: 'saved', type: 'content'});
 	myFtClient.init(clientOpts);
 
-	myFtUi.init({anonymous: !(/FTSession=/.test(document.cookie))});
+	myFtUi.init({anonymous: !!(/FTSession=/.test(document.cookie))});
```
Comment (andygnewman) [inline client/main.js]: why do we have an extra ! in here?

File: client/main.js [hunk 8]
```diff
@@ -27,6 +27,13 @@ require('next-js-setup').bootstrap(({flags}) => {
 	const share = require('./components/share/main');
 	const labsShare = require('./components/labsshare/main');
 	const trackEvent = require('./components/utils/tracking');
+
+	const ftlabsAdBlockerHandling = require('./components/ftlabsadblockerhandling/main');
+
+	if(flags.get('ftlabsAdBlockerHandling') === 1){
```
Comment (andygnewman) [inline client/main.js]: shouldn't need the === 1

```
	if(flags.get('ftlabsAdBlockerHandling')){
		ftlabsAdBlockerHandling.init();
	}
```
should do it

File: views/partials/ads/banlb.html [hunk 9]
```diff
@@ -1,5 +1,5 @@
 <div
-	class="article-header-advert o-ads o-ads__center"
+	class="above-header-advert o-ads o-ads__center"
```
Comment (andygnewman) [inline views/partials/ads/banlb.html]: why do we have changes here?

File: views/partials/ads/top.html [hunk 10]
```diff
@@ -1,12 +1,16 @@
 <div
-	class="article-header-advert o-ads o-ads__center"
+	{{#if @root.flags.adLoadingImprovements}}
```
Comment (andygnewman) [inline views/partials/ads/top.html]: why do we have changes here?

File: client/components/ftlabsadblockerhandling/main.js [hunk 11]
```diff
@@ -0,0 +1,71 @@
+const approaches = [require('./no-vowels'), require('./bart'), require('./ad-revenue'), require('./lets-talk'), require('./ad-blocking-articles')];
```
Comment (andygnewman) [inline client/components/ftlabsadblockerhandling/main.js]: This is getting long would be better to reformat;
```
const approaches = [
    require('./no-vowels'),
    require('./bart'),
    require('./ad-revenue'),
    require('./lets-talk'),
    require('./ad-blocking-articles')
];
````

**Comment by andygnewman** [comment]
👍

**Comment by i-like-robots** [comment]
This appears to have added ~340kb to our JS bundle.

_very large frowning face_

**Comment by andygnewman** [comment]
mmm, will take a look at what we can do to reduce. Any tips welcome.

**Comment by i-like-robots** [comment]
It looks like it may not be this... I will investigate and try to pin down when this jump happened.

**Pull request merged**